RecoBTag/ONNXRuntime/plugins/ParticleNetSonicJetTagsProducer.cc

```diff
@@ -0,0 +1,317 @@
+                                         "probHqqqq",
+                                         "probQCDbb",
+                                         "probQCDcc",
+                                         "probQCDb",
+                                         "probQCDc",
+                                         "probQCDothers",
+                                     });
+  desc.addOptionalUntracked<bool>("debugMode", false);
+
+  descriptions.addWithDefaultLabel(desc);
+}
+
+void ParticleNetSonicJetTagsProducer::produce(edm::Event &iEvent,
+                                              const edm::EventSetup &iSetup,
+                                              Output const &iOutput) {
+  edm::Handle<TagInfoCollection> tag_infos;
+  iEvent.getByToken(src_, tag_infos);
+
+  // initialize output collection
+  std::vector<std::unique_ptr<JetTagCollection>> output_tags;
+  if (!tag_infos->empty()) {
+    auto jet_ref = tag_infos->begin()->jet();
+    auto ref2prod = edm::makeRefToBaseProdFrom(jet_ref, iEvent);
+    for (std::size_t i = 0; i < flav_names_.size(); i++) {
+      output_tags.emplace_back(std::make_unique<JetTagCollection>(ref2prod));
+    }
+  } else {
+    for (std::size_t i = 0; i < flav_names_.size(); i++) {
+      output_tags.emplace_back(std::make_unique<JetTagCollection>());
+    }
+  }
+  for (unsigned jet_n = 0; jet_n < tag_infos->size(); ++jet_n) {
+    const auto &taginfo = (*tag_infos)[jet_n];
+    std::vector<float> outputs(flav_names_.size(), 0);  // init as all zeros
+    if (!taginfo.features().empty() && jet_n == 0) {
+      // run prediction and get outputs
+      const auto &output1 = iOutput.begin()->second;
+      const auto &outputs_from_server = output1.fromServer<float>()[0];
+      std::copy(outputs_from_server.begin(), outputs_from_server.end(), outputs.begin());
+    }
+
+    const auto &jet_ref = tag_infos->at(jet_n).jet();
+    for (std::size_t flav_n = 0; flav_n < flav_names_.size(); flav_n++) {
+      (*(output_tags[flav_n]))[jet_ref] = outputs[flav_n];
+    }
+  }
+
+  if (debug_) {
+    std::cout << "=== " << iEvent.id().run() << ":" << iEvent.id().luminosityBlock() << ":" << iEvent.id().event()
+              << " ===" << std::endl;
+    for (unsigned jet_n = 0; jet_n < tag_infos->size(); ++jet_n) {
+      const auto &jet_ref = tag_infos->at(jet_n).jet();
+      std::cout << " - Jet #" << jet_n << ", pt=" << jet_ref->pt() << ", eta=" << jet_ref->eta()
+                << ", phi=" << jet_ref->phi() << std::endl;
+      for (std::size_t flav_n = 0; flav_n < flav_names_.size(); ++flav_n) {
+        std::cout << "    " << flav_names_.at(flav_n) << " = " << (*(output_tags.at(flav_n)))[jet_ref] << std::endl;
+      }
+    }
+  }
+
+  // put into the event
+  for (std::size_t flav_n = 0; flav_n < flav_names_.size(); ++flav_n) {
+    iEvent.put(std::move(output_tags[flav_n]), flav_names_[flav_n]);
+  }
+}
+
+std::vector<float> ParticleNetSonicJetTagsProducer::center_norm_pad(const std::vector<float> &input,
```

Comment:
this could also be abstracted into a common helper function to be shared with the non-SONIC version

Reply:
This was moved to deep_helpers, and BoostedJetONNXJetTagsProducer.cc uses this new version now too